### PR TITLE
core: add argument and result types getters to CallableOpInterface

### PR DESCRIPTION
**Pull request by superlopuh**
This adds some more of the capabilities of the MLIR equivalent.

Note stacked PR

### CODE REVIEW

**Comment by Moxinilian** [comment]
## Codecov Report

This pull request has no tests Mr Lopoukhine.

> Project coverage is actually pretty good. Comparing base [(c366844)](https://app.codecov.io/gh/xdslproject/xdsl/commit/c3668445f8e4887b26d8ce40e2fa67ae5c6c6ad8?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=xdslproject) to head [(525ddda)](https://app.codecov.io/gh/xdslproject/xdsl/pull/2535?dropdown=coverage&src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=xdslproject).

<details><summary>Additional details and impacted files</summary>

actually no thanks</details>

[:umbrella: View full report in Codecov by Théo](https://youtu.be/dQw4w9WgXcQ)
:loudspeaker: Have feedback on the report? Yeah that makes sense.

**Comment by superlopuh** [comment]
Weird indeed, I was aware that there were no tests but the fact that CodeCov doesn't bring this up is somewhat surprising. I'll add some

**Review by AntonLydike — APPROVED** [review]
Really nice! I like it! Thanks! :rocket:

File: xdsl/traits.py [hunk 1]
```diff
@@ -413,11 +413,22 @@ class CallableOpInterface(OpTrait, abc.ABC):
...
+
+    @classmethod
+    @abc.abstractmethod
+    def get_argument_types(cls, op: Operation) -> tuple[Attribute, ...]:
+        raise NotImplementedError()
+
+    @classmethod
+    @abc.abstractmethod
+    def get_result_types(cls, op: Operation) -> tuple[Attribute, ...]:
+        raise NotImplementedError()
```
Comment (JosseVanDelm) [inline xdsl/traits.py:431]: Is there a reference you can give to the MLIR equivalent?

**Review by JosseVanDelm — APPROVED** [review]
Looks good, I just noticed that I don't understand a single thing about MLIR interfaces defined in tablegen

File: xdsl/traits.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/traits.py:431]: I tried to hunt for this, but the best one I found was at the link in the docstring for the whole trait. I also found it quite difficult to navigate the tablegen.

**Pull request merged**